### PR TITLE
SimpleCounter uses a HashMap by default and exposes a method to

### PR DESCRIPTION
**Pull request by nh13**
sub-classes to make the internal counting object.

@tfenne you mentioned you hit a case where you were creating many counters and the reflection code was the bottleneck.  I also hit such a case.  Take a look.

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/util/NumericCounter.scala [hunk 1]
```diff
@@ -42,9 +41,13 @@ object NumericCounter {
   * zero for any item that has not been counted yet.  Implements some useful methods
   * to compute statistics when the objects being counted are numeric types.
   */
-class NumericCounter[T](implicit numeric: Numeric[T], tt: TypeTag[T]) extends SimpleCounter[T] {
+class NumericCounter[T](implicit numeric: Numeric[T]) extends SimpleCounter[T] {
+  import scala.collection.JavaConversions._
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/NumericCounter.scala]: Could you import just the 1 or 2 conversions that are needed?  I dislike importing all of them as they can have unintended consequences.

File: src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala [hunk 2]
```diff
@@ -42,19 +40,14 @@ object SimpleCounter {
...
-      new java.util.TreeMap[T,Long]()
-    }
-    else {
-      new util.HashMap[T,Long]()
-    }
-  }
+  /** Creates the map in which we store the counts. */
+  protected def makeCounts(): scala.collection.mutable.Map[T, Long] = new util.HashMap[T,Long]()
+
+  private val counts = makeCounts()
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala]: Can you add a type annotation here just to be super-clear?

File: src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala [hunk 3]
```diff
@@ -42,19 +40,14 @@ object SimpleCounter {
...
-    val clazzComparable = ReflectionUtil.typeTagToClass[Comparable[T]]
-    if (clazzComparable.isAssignableFrom(clazzT)) {
-      new java.util.TreeMap[T,Long]()
-    }
-    else {
-      new util.HashMap[T,Long]()
-    }
-  }
+  /** Creates the map in which we store the counts. */
+  protected def makeCounts(): scala.collection.mutable.Map[T, Long] = new util.HashMap[T,Long]()
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala]: `makeMap`?

File: src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala [hunk 4]
```diff
@@ -69,6 +62,6 @@ class SimpleCounter[T](implicit tt: TypeTag[T]) extends Iterable[(T, Long)] {
   /** Iterates over all the objects and their counts. */
   override def iterator: Iterator[(T, Long)] = this.counts.entrySet().iterator().map(entry => (entry.getKey, entry.getValue))
 
-  /** Gets sum of counts. */
-  def totalCounts: Long = _totalCounts.toInt
+  /** Gets the number of counts stored in this counter (sum over all count). */
+  def sumOfCounts: Long = _totalCounts.toInt
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala]: how about `totalObservations`?  Or just `total`?  I think the problem with these kinds of methods is finding good names.

File: src/main/scala/com/fulcrumgenomics/util/NumericCounter.scala [hunk 5]
```diff
@@ -112,5 +115,8 @@ class NumericCounter[T](implicit numeric: Numeric[T], tt: TypeTag[T]) extends Si
       }
     }
   }
+
+  /** Gets sum of values stored in this counter (sum over all (count * value)). */
+  def sumOfValues: T = totalSum
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/NumericCounter.scala]: `totalMass`?

**Comment by tfenne** [comment]
@nh13 Naming is hard :(  A few minor comments.

File: src/main/scala/com/fulcrumgenomics/util/NumericCounter.scala (hunk 1 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/util/NumericCounter.scala]: Done.

File: src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala (hunk 2 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala]: Done

File: src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala (hunk 3 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala]: Done.

File: src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala (hunk 4 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/util/SimpleCounter.scala]: `total`

**Pull request merged**